server/src/main/java/io/crate/execution/dml/TranslogWriter.java

```diff
@@ -0,0 +1,89 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.dml;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+
+import io.crate.sql.tree.BitString;
+
+/**
+ * Converts the translog as represented by a Map to a BytesReference
+ */
+public class TranslogWriter {
+
+    private TranslogWriter() { }
+
+    public static BytesReference convert(Map<String, Object> content) {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            try (
+                XContentBuilder builder = XContentFactory.json(output)) {
+                builder.startObject();
+                for (var entry : content.entrySet()) {
+                    writeValue(entry.getKey(), entry.getValue(), builder);
+                }
+                builder.endObject();
+            }
+            // XContentBuilder must be closed before we access the bytes
+            return output.bytes();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static void writeValue(String key, Object value, XContentBuilder builder) throws IOException {
+        builder.field(key);
+        switch (value) {
+            case null -> writeValue(value, builder);
+            case List<?> l -> {
+                builder.startArray();
+                for (var o : l) {
+                    writeValue(o, builder);
+                }
+                builder.endArray();
+            }
+            case Map<?, ?> m -> {
+                builder.startObject();
+                for (var entry : m.entrySet()) {
+                    writeValue(entry.getKey().toString(), entry.getValue(), builder);
+                }
+                builder.endObject();
+            }
+            case Object _ -> writeValue(value, builder);
+        }
+    }
+
+    private static void writeValue(Object value, XContentBuilder builder) throws IOException {
+        switch (value) {
+            case null -> builder.nullValue();
+            case BitString bitString -> builder.value(bitString.bitSet().toByteArray());
+            case Object o -> builder.value(o);
+        }
+    }
```

Comment:
That we loose the innate type information that the value indexers have and need to work with casts / spread type representation knowledge is also a bit of a concern

Reply:
I was thinking along the same lines, and was surprised that we only needed to specialise BitString here.  There are tests in DataTypeTestCase that check the translog roundtrip, and this was the only one that didn't work immediately.  We end up preserving all the type information because types are tied to the Schema, and so SourceParser knows how to cast the basic representation in the json to whatever the DataType wants.